GSoC Proposal for project - Implement Bindings in Go

srinivasyadav18 opened the pull request:
Please review it. Im Happy to receive any kind of feedback . Thank You 😊

srinivasyadav18 committed:
GSoc Proposal for project Implements Bindings in Go

jaylett commented on proposal.rst:
```diff
@@ -67,27 +67,28 @@
 
 ======================================
-FILLME WITH THE TITLE OF YOUR PROPOSAL
+GSOC 2020 - PROPOSAL
```
The title of your proposal should go here.

jaylett commented on proposal.rst:
```diff
@@ -162,7 +179,7 @@ FILLME
 .. Code to make for a worthwhile project.
 
-FILLME
+Monday–Saturday 5am-4pm UTC 
```
I'd suggest that 11 hours a day six days a week is not really sustainable for the entire length of the programme. While this is obviously up to you, please take seriously our guidance notes in the comment above — because if you burn out in the first few weeks by pushing yourself too hard, that will almost certainly result in you being failed at one of the later checkpoints (because you'll be unable to contribute effectively after you've burned out). You can generally get as much or more done sustainably (so over an indefinite period) by keeping your regular hours down to more like 40 hours a week. While we don't want to discourage your enthusiasm, we see this as a valuable part of the programme: learning how to contribute in a sustainable fashion, whether to open source or to other kinds of jobs.

jaylett commented on proposal.rst:
```diff
@@ -171,24 +188,27 @@ organisation(s)?**
 
-Your Project
-============
+Implement Bindings for Go
```
We meant you to put the project title at the top of the document rather than here. It's not a huge problem though!

jaylett commented on proposal.rst:
```diff
@@ -171,24 +188,27 @@ organisation(s)?**
 **Why have you chosen this particular project?**
 
-FILLME
+My main areas of interest includes C++ and Go. They provide greater flexibility and concurrency.
```
It isn't hugely important, but if you make a statement like "X provides greater Y", as you do here, this is meaningless without something to compare _to_. "X provides greater Y than Z" makes sense. It's almost never true that X provides greater Y than _everything else it could be compared to_, simply because of the enormous range of technologies out there. (Also, "concurrency" is a tricky beast. Do you mean concurrency within a single process, using threads and fibers; concurrency within a single machine or container which might also use multiple processes; concurrency within multiple machines in a cluster located closely within the network and/oir hardware topology; concurrency across arbitrary machines on the internet; and so on. I'm sure you know what you mean here, but it's always worth taking time to step back and think about whether you've provided enough detail to communicate that.)

A specific reason I'm pointing this out is that "<my favourite thing> provides greater <something or another>" is commonly seen online in people who get very close to one particular set of technologies, and believe they're better than everything else — but don't stay sufficiently up to date with other technologies to be able to know that for sure. So these kinds of statements online tend to be disregarded as being more about the person making them than about technology. It's good practice therefore to avoid statements like this, and instead be more precise about what you mean.

jaylett commented on proposal.rst:
```diff
@@ -267,8 +292,69 @@ Project Timeline
+Each week for first 3 days I would work on provding API , that is writing swig interface 
+for the functions and classes which aren't generated from 
+existing swig interfaces(xapian-headers.i and xapian-head.i) with typemaps and other features of swig in go.i.
+Next 2 days documenting the code with examples and the last day for testing.
```
When you talk about testing here, do you mean automated tests?

jaylett commented on proposal.rst:
```diff
@@ -267,8 +292,69 @@ Project Timeline
+Before May 18,2020 :
+I would finish up the build system part of work for go bindings that 
+are generated purely from xapian-headers.i and xapian-head.i such that
+these can be used like any other bindings and in any platform that other xapian bindings can be used in.
```
A word of warning: unless you have access to a Windows machine (or VM) it may be difficult to get this working. There's a good chance that the go build system will have wrinkles specific Windows. What I'd suggest is that you move that to a stretch goal to the overall project. (Fully-functional go bindings that work on linux alone is better than partially-functional bindings that work on linux, Windows & macos.)

ojwb replied:
It would certainly be good to get them working portably, but I'd agree that getting them working well on one platform first is probably a better approach.  Windows and macOS both have a number of differences from Unix-like platforms that can suck up a lot of time.

jaylett commented on proposal.rst:
```diff
@@ -267,8 +292,69 @@ Project Timeline
+First Month : 
+
+  Week1 :
+        Work on Document , Registry , Query , Stem implementation , Stem classes.
```
You've listed a bunch of classes in each week, but it's not clear what you intend to do for them. Generally there are going to be several things you need to do for each class:

1. Wrapping anything that SWIG doesn't do automatically. This may involve typemaps, or may involve more complex custom code and/or additional SWIG directives.
2. Making the resultant API more idiomatic for golang users. Typically the main wins are in iterators, but there may be other things that better match go if they are abstracted further away from the C++ API.
3. Making sure exceptions work as intended. golang doesn't have exceptions; instead it has a return code pattern.
4. Supporting Directors for the class if necessary (which is a SWIG concept allowing you to subclass a C++ class in the target language). This only applies to some classes; most of Xapian you don't need or want to subclass. (Document, for instance, isn't helpful to subclass. MatchSpy is.)

What I think you should do as much as possible this week is to look through the different classes you've listed, and figure out which of the things above make sense to do. Then you may want to think in terms of functionality to deliver piece by piece. For instance, you might want to first get to a place where you can write a simple query system in golang is probably helpful. For that, you need at least Database, Document, Enquire, Query, QueryParser, MSet and MSetIterator, and probably Stem (but you're unlikely to need StemImplementation) — possibly with some methods ignored. Then you could add basic indexing (WritableDatabase and TermGenerator), and then gradually fill out the other functionality (which will depend on what you think is most useful — perhaps you want to add support for Stopper, or for the other iterators, or for configuring the weighting scheme used when querying).

A good way of figuring out the chunks of functionality would be to run through the getting started / user guide, which introduces different classes and functionality in a sensible order.

ojwb replied:
There's probably nothing needed over the default SWIG wrapping for some classes.  For example, there's no special handling for `Xapian::Registry` in any of the existing bindings, so I doubt any would be needed for Go (try `git grep -w Registry` in xapian-bindings to see this).

jaylett commented on proposal.rst:
```diff
@@ -267,8 +292,69 @@ Project Timeline
+  Week9 :
+        Work on Database class.
+  Week10 :
+        Work on Exceptions.
```
The work in these weeks (9 & 10) are almost certainly more important than PostingSource, MatchSpy, and some of the others you have in the second month. However you might be able to split the work here — ensure enough of the Database functionality is working for most uses, then move on to things like MatchSpy before coming back to tidy up the rest of Database. The more you can get into this level of detail, the easier it will be to make a success of the project — so we expect quite a lot of detail in the proposal.

jaylett reviewed (changes requested):
The main thing I'd suggest is to get more into the detail of what you need to do. When you're talking about working on a class like Document, what will you need to do? What does SWIG already give you for free? Will you be adding new golang code, new C++ code, new typemaps, or something else?

It's probably worth looking at this project description and plan: https://trac.xapian.org/wiki/GSoC2011/SupportLua/ProjectPlan

It was from a proposal a number of years ago for adding Lua support using SWIG. It was a successful project, although I'd have preferred to see more detail in the timeline / project plan at the end, in the way I've indicated in some of my comments here.

ojwb commented:
> It was from a proposal a number of years ago for adding Lua support using SWIG. It was a successful project, although I'd have preferred to see more detail in the timeline / project plan at the end, in the way I've indicated in some of my comments here.

It's probably worth noting that Xiaona Han had already managed to get basic Lua bindings working from scratch before we were reviewing proposals, so we had a lot of confidence they could make a success of the project.

srinivasyadav18 committed:
Revised proposal

srinivasyadav18 committed:
Fix Date

srinivasyadav18 committed:
Revised plan

ojwb commented on proposal.rst:
```diff
@@ -351,8 +352,8 @@ Plan :
-        for i := range container.Iter(){
-          i.GetData() // methods to get information from the iterator at that position.
+        for i := range doc.terms(){
+          i.GetTerm() // methods to get term from the iterator at that position.
```
This sounds like a good way to wrap these, but what's actually involved in doing so?

srinivasyadav18 replied:
you mean to show me how its done internally in go?...If yes then ill code it and show...thats done using channels in go ... in a brief range gets data from the channel which is open until iterator is reaches the last element , then channel closes and range stops..

srinivasyadav18 replied:
> This sounds like a good way to wrap these, but what's actually involved in doing so?

Hey I have added a link how this can be done with small piece of code.

ojwb replied:
Kind of, but that's just showing how to create a channel using only Go code.  The real (and perhaps more tricky, though perhaps not) question is what's actually involved in doing the same but with data coming from a Xapian iterator wrapped via SWIG.

srinivasyadav18 replied:
> Kind of, but that's just showing how to create a channel using only Go code. The real (and perhaps more tricky, though perhaps not) question is what's actually involved in doing the same but with data coming from a Xapian iterator wrapped via SWIG.

Hey I have put the code for implementing termiterator in document class (I added only raw code how termiterator is implemented but the other parts of document interface needs rewrapping which will be done week 2 , so please consider only for-range construct  implementation part)

srinivasyadav18 replied:
> Kind of, but that's just showing how to create a channel using only Go code. The real (and perhaps more tricky, though perhaps not) question is what's actually involved in doing the same but with data coming from a Xapian iterator wrapped via SWIG.

Hey , I have put the code for iterators , errors , type conversions , rewrapping . Please review it. Thank You 😊

ojwb replied:
> https://github.com/srinivasyadav18/xapian-gsoc-plan/blob/master/go.i#L43

This looks promising, though it seems only provides access to the term name.

What about the other accessor methods on the `TermIterator` object?  At least for some of them, they can involve significant work so just reading all of them up front isn't great.  The Python bindings have attempted to address such issues already - maybe that's worth looking at to see if the approach works for Go too.

It's also good to try to partly automate such wrapping via SWIG macros - otherwise you're going to end with a lot of code cut-and-pasted and then modified to support all the different methods which return a `TermIterator`, and that's not great for maintainability.

srinivasyadav18 replied:
Hey that’s not a big issue ..I could just send the term iterator object over the channel ..so now through iteration we can access all the method of it.

srinivasyadav18 committed:
iterators using channels

srinivasyadav18 committed:
TermIterator in Document class 

Sample implementation of termiterator in document class using for-range construct

srinivasyadav18 committed:
Revised Proposal

srinivasyadav18 reviewed (commented):
Thank You for the feedback. I have made the changes to proposal please review it.

jaylett commented on proposal.rst:
```diff
@@ -267,8 +304,191 @@ Project Timeline
+        begin and end iterators in one function call as below.
+        /* begin,end := doc.Terms() */ 
+
+      * Go supports errors as return values . A language like c++ have try catch block Go has three constructs for dealing
```
You talk about handling exceptions using panic/defer/recover, but is this the right approach? Should our exceptions instead be converted into return values? Can you write some example golang showing how to handle a common case? (The most obvious is probably when the `Database()` raises an exception if you tell it to open a database that doesn't exist.)

srinivasyadav18 replied:
> You talk about handling exceptions using panic/defer/recover, but is this the right approach? Should our exceptions instead be converted into return values? Can you write some example golang showing how to handle a common case? (The most obvious is probably when the `Database()` raises an exception if you tell it to open a database that doesn't exist.)

Hey I have put the example case for Database already...please check, it catches the error as return value during call of the constructer. I have followed the pattern that go  standard library follows.

srinivasyadav18 committed:
Add URLs

srinivasyadav18 committed:
Go std lib error pattern

srinivasyadav18 committed:
Iterators and Errors in main.go

ojwb commented on proposal.rst:
```diff
@@ -369,6 +369,9 @@ Plan :
         with exceptions, they are panic defer and recover.A Panic is similar to an exception which can occur an runtime exception.
         C++ exceptions can be handled in go from swig wrappers as follows(https://github.com/srinivasyadav18/xapian-gsoc-plan/blob/master/example.i#L16).
         Which ever class function throws an exception in c++ , the wrapped function in Go returns the error as value.
+        Database error handling - https://github.com/srinivasyadav18/xapian-gsoc-plan/blob/master/go.i#L77
```
We really need to find a way to automate such error handling - having to write this by hand for every wrapped method will make the Go bindings too hard to maintain going forwards.

For a typical new C++ method, we get automatic wrapping for all the bindings for **zero** extra work.  In some cases, extra work to get a new C++ method wrapped is needed, but it's rare.

Having to do something for every new wrapped method scales badly in terms of developer effort, and we definitely need to avoid that.  We've had non-SWIG-based bindings which required manual update before (XS bindings for Perl and JNI bindings for Java) and they constantly lagged behind the C++ API, and that frustrated both users and developers.

srinivasyadav18 replied:
> We really need to find a way to automate such error handling - having to write this by hand for every wrapped method will make the Go bindings too hard to maintain going forwards.
> 
> For a typical new C++ method, we get automatic wrapping for all the bindings for **zero** extra work. In some cases, extra work to get a new C++ method wrapped is needed, but it's rare.
> 
> Having to do something for every new wrapped method scales badly in terms of developer effort, and we definitely need to avoid that. We've had non-SWIG-based bindings which required manual update before (XS bindings for Perl and JNI bindings for Java) and they constantly lagged behind the C++ API, and that frustrated both users and developers.

Since errors are treated as values in Golang I am explicitly hand wrapping for the function that throws exception. I would work on it and find a way to automate for wrapping error  in community bonding period.

srinivasyadav18 replied:
> We really need to find a way to automate such error handling - having to write this by hand for every wrapped method will make the Go bindings too hard to maintain going forwards.
> 
> For a typical new C++ method, we get automatic wrapping for all the bindings for **zero** extra work. In some cases, extra work to get a new C++ method wrapped is needed, but it's rare.
> 
> Having to do something for every new wrapped method scales badly in terms of developer effort, and we definitely need to avoid that. We've had non-SWIG-based bindings which required manual update before (XS bindings for Perl and JNI bindings for Java) and they constantly lagged behind the C++ API, and that frustrated both users and developers.

Hey do you mean to say if a new c++ method gets added then that should get rewrapped automatically instead of hand wrapping?if that's the case..I think most of the work is in writing a script which generates the wrappers automatically. You are correct, we get automatic wrapping for other bindings with no extra work, but the swig support for Go doesn't produce the API which a go user can use naturally
. So this means that we need to provide second level of wrapper (kind of another swig software ) to generate wrappers for swig generated objects to produce a simple API.

srinivasyadav18 replied:
Could you please explain the comment in detail..I didn't understand completely

ojwb replied:
> I think most of the work is in writing a script which generates the wrappers automatically

We already have a tool which generates wrappers automatically - it's called "SWIG".

You should be using SWIG to achieve this, not writing extra layers of scripts above or around it.  SWIG is up for the job of automating the majority of the wrapping for all the other languages we use it for - I don't see why Go is different in that regard.

ojwb commented on proposal.rst:
```diff
@@ -369,6 +369,9 @@ Plan :
         Which ever class function throws an exception in c++ , the wrapped function in Go returns the error as value.
+        Database error handling - https://github.com/srinivasyadav18/xapian-gsoc-plan/blob/master/go.i#L77
+        https://github.com/srinivasyadav18/xapian-gsoc-plan/blob/master/main.go#L36
+        error (pastebin) - https://pastebin.com/AuFpiRdQ
```
Please include content in your proposal rather than linking to external sites.  The final proposal you submit is essentially what you're saying you'll do if chosen, so it ought to be self-contained rather referring out to information on external sites can change or even disappear (content on paste sites typically expires after a while so is particular poor to refer to).

It would also be better to inline or include the content referred to in https://github.com/srinivasyadav18/xapian-gsoc-plan/blob/master/main.go for the same reasons.  That's at least that's under version control, so not quite as bad as linking to a paste site.

srinivasyadav18 replied:
> Please include content in your proposal rather than linking to external sites. The final proposal you submit is essentially what you're saying you'll do if chosen, so it ought to be self-contained rather referring out to information on external sites can change or even disappear (content on paste sites typically expires after a while so is particular poor to refer to).
> 
> It would also be better to inline or include the content referred to in https://github.com/srinivasyadav18/xapian-gsoc-plan/blob/master/main.go for the same reasons. That's at least that's under version control, so not quite as bad as linking to a paste site.

Is it fine the proposal will get longer? I have include the code now in proposal . If it is not good please give a feedback. I have put even the links at the end in references.

ojwb replied:
Don't worry about length (at least if the content is relevant as this is).

srinivasyadav18 committed:
include code in proposal

srinivasyadav18 committed:
include code in proposal

srinivasyadav18 committed:
Revised proposal

srinivasyadav18 committed:
Revised proposal

srinivasyadav18 committed:
Revised proposal

srinivasyadav18 committed:
Revised proposal

srinivasyadav18 committed:
Revised proposal

ojwb commented:
Note you'll need to submit a "final" PDF version of this at https://summerofcode.withgoogle.com/ to be eligible for GSoC - currently I only see a draft version there, and Google require a final version before the deadline or else we can't select you.

You should be able to generate a suitable pdf from the .rst version using make pdf.

You can update the "final" version as many times as you like up to the deadline, so it's wise to submit early and not run the risk of internet problems preventing you submitting in time.

ojwb commented:
Thanks for you proposal.

The purpose of this PR was to provide feedback on the draft proposal rather than actually as something to merge into the template, so closing as the proposal deadline has now passed.
